[WIP] moose-test auto scaling

lindsayad opened the pull request:
Converts the moose test suite over to using automatic scaling by default.

This process unearthed a coupled of bugs with respect to time integration.

lindsayad committed:
Make moose-test work with automatic scaling

Refs #12601

lindsayad committed:
Ensure scaling factors done right for Crank-Nicolson

Variable scaling factors must be calculated before the residual
compute call done in CrankNicolson::init

lindsayad committed:
initial -> scaling

lindsayad committed:
Don't use scaling Jacobian for true Jacobian in explicit Euler

lindsayad committed:
Issue warnings in cases where auto scaling cant be conducted

Some cases:
  - PETSc is not available
  - The user is using a non-PETSc matrix
  - There is no matrix but the user has explicitly requested auto
    scaling

lindsayad committed:
Refactor where scaling data lives and is computed from

This commit moves calculation of the scaling Jacobian from the
beginning of NonlinearSystem::solve to PicardSolve::solve. This
ensures that the Picard norm will match the initial non-linear
residual norm of the master application. Additionally all scaling
data members were moved from FEProblemBase and NonlinearSystem
into NonlinearSystemBase. It just seems appropriate that the
nonlinear system (where traditional nonlinear or eigen in the future)
owns the data related to scaling of the residual and Jacobian.
FEProblemBase to me is an object that should be designed to hold
the data related to using finite elements to create an algebraic
system. The nonlinear system then solves that system using Newton
technology (typically)

lindsayad committed:
Move computingScalingJacobian information to SystemBase

This enables:
- Easily setting whether we are computing the scaling jacobian for
  the displaced nonlinear system
- Checking in a kernel whether we are computing the scaling jacobian
  regardless of whether we are in the regular or displaced nonlinear
  system

lindsayad committed:
Dont scale anytime we have JFNK

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/362058/) on ccbf24c : invalidated by @lindsayad

permcody commented on test/src/base/MooseTestApp.C:
```diff
@@ -31,7 +31,7 @@ validParams<MooseTestApp>()
                                    false,
                                    "Don't register test objects and syntax");
-  params.set<bool>("automatic_automatic_scaling") = false;
+  params.set<bool>("automatic_automatic_scaling") = true;
```
![image](https://user-images.githubusercontent.com/3036960/60897715-ca1e8c00-a225-11e9-8bde-403bd683307a.png)

BTW - How does this work? I didn't see where you are retrieving this parameter. Is it a typo?

permcody replied:
Oh I see - it's already in the codebase, just not this PR.

moosebuild commented:
All [jobs](https://civet.inl.gov/event/43792/) on ccbf24c : invalidated by @lindsayad

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/362065/) on ccbf24c wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/13687/site)

This comment will be updated on new commits.

tophmatthews commented:
ping @gardnerru

lindsayad committed:
Address failing recover test

How to deal with this test? If the user has requested automatic
scaling once, then we really should not be scaling upon recovery.
We should use the factors from the previous portion of the sim.
In order to do that, the scaling factors need to be restartable
data. however, in order for the data to be restartable,
MooseVariableBase needs to inherit from MooseObject. So maybe
now is finally the time to make MooseVariableBase inherit from
MooseObject???

moosebuild commented:
All [jobs](https://civet.inl.gov/event/44230/) on 6c8d620 : canceled by @lindsayad

Currently a WIP

tophmatthews commented:
@lindsayad This PR seems to take scaling to a more appropriate level of vetting, are there plans to get it in, and move down the line to other modules? I haven't had time to really start applying this and mortar to bison calcs, but plan to in the next couple of weeks.

lindsayad commented:
Yes, I do plan to get it in. But there is at least one failing test that I
want to fix up using https://github.com/idaholab/moose/pull/13769, so this
has to wait until I'm done with that PR

On Thu, Jul 25, 2019 at 7:54 AM Topher Matthews <notifications@github.com>
wrote:

> @lindsayad <https://github.com/lindsayad> This PR seems to take scaling
> to a more appropriate level of vetting, are there plans to get it in, and
> move down the line to other modules? I haven't had time to really start
> applying this and mortar to bison calcs, but plan to in the next couple of
> weeks.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/13687?email_source=notifications&email_token=ACOGA4HWIRODQGONGMJBAUDQBG5BFA5CNFSM4H5BWKT2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD2ZXJMI#issuecomment-515077297>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACOGA4GJYX6B7CAIOUCFL6TQBG5BFANCNFSM4H5BWKTQ>
> .
>

tophmatthews commented:
Great! Just wanted to make sure the stalebot dosen't bite this to oblivion

tophmatthews commented:
> Great! Just wanted to make sure the stalebot dosen't bite this to oblivion

💀

tophmatthews commented:
What happened with this one @lindsayad. Hopefully it’s redundant now, but I’m worried about the “unearthed a couple of bugs”

lindsayad commented:
If all you use is the default time integrator then there are no concerns

> On Sep 17, 2019, at 4:33 PM, Topher Matthews <notifications@github.com> wrote:
> 
> What happened with this one @lindsayad. Hopefully it’s redundant now, but I’m worried about the “unearthed a couple of bugs”
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

lindsayad commented:
Keeping this as a reminder. A lot of the framework work from this is already in, but it would be nice to get all the tests working (which this does except for one at last check). Most important thing that remains is to do convergence studies with automatic scaling for all the different time integrators

lindsayad commented:
Still on my todo list...

permcody commented:
Uh huh.... sure....

On Sun, Nov 17, 2019 at 1:17 PM Alex Lindsay <notifications@github.com>
wrote:

> Still on my todo list...
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/13687?email_source=notifications&email_token=AAXFOIF7FR5H3LH6KZM3UQLQUGDBZA5CNFSM4H5BWKT2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEEISJ5I#issuecomment-554771701>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAXFOIEAXNTVA6IJKHTMHYTQUGDBZANCNFSM4H5BWKTQ>
> .
>

lindsayad commented:
It's 90% done

On Sun, Nov 17, 2019 at 2:52 PM Cody Permann <notifications@github.com>
wrote:

> Uh huh.... sure....
>
> On Sun, Nov 17, 2019 at 1:17 PM Alex Lindsay <notifications@github.com>
> wrote:
>
> > Still on my todo list...
> >
> > —
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/idaholab/moose/pull/13687?email_source=notifications&email_token=AAXFOIF7FR5H3LH6KZM3UQLQUGDBZA5CNFSM4H5BWKT2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEEISJ5I#issuecomment-554771701
> >,
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AAXFOIEAXNTVA6IJKHTMHYTQUGDBZANCNFSM4H5BWKTQ
> >
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/13687?email_source=notifications&email_token=ACOGA4BLCEXLOCA7YL6ZC7DQUHDMDA5CNFSM4H5BWKT2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEEIYP6I#issuecomment-554797049>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACOGA4DAAGEWUHON5SI7DSLQUHDMDANCNFSM4H5BWKTQ>
> .
>

permcody commented:
Sounds like something @Derek Gaston <friedmud@gmail.com> would say.

On Sun, Nov 17, 2019 at 6:01 PM Alex Lindsay <notifications@github.com>
wrote:

> It's 90% done
>
> On Sun, Nov 17, 2019 at 2:52 PM Cody Permann <notifications@github.com>
> wrote:
>
> > Uh huh.... sure....
> >
> > On Sun, Nov 17, 2019 at 1:17 PM Alex Lindsay <notifications@github.com>
> > wrote:
> >
> > > Still on my todo list...
> > >
> > > —
> > > You are receiving this because you commented.
> > > Reply to this email directly, view it on GitHub
> > > <
> >
> https://github.com/idaholab/moose/pull/13687?email_source=notifications&email_token=AAXFOIF7FR5H3LH6KZM3UQLQUGDBZA5CNFSM4H5BWKT2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEEISJ5I#issuecomment-554771701
> > >,
> > > or unsubscribe
> > > <
> >
> https://github.com/notifications/unsubscribe-auth/AAXFOIEAXNTVA6IJKHTMHYTQUGDBZANCNFSM4H5BWKTQ
> > >
> > > .
> > >
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/idaholab/moose/pull/13687?email_source=notifications&email_token=ACOGA4BLCEXLOCA7YL6ZC7DQUHDMDA5CNFSM4H5BWKT2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEEIYP6I#issuecomment-554797049
> >,
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/ACOGA4DAAGEWUHON5SI7DSLQUHDMDANCNFSM4H5BWKTQ
> >
> > .
> >
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/13687?email_source=notifications&email_token=AAXFOIAOCAXNHPS27OSZ4KDQUHEMBA5CNFSM4H5BWKT2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEEIYWKQ#issuecomment-554797866>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAXFOIBVIWCYLGGCIYTEKULQUHEMBANCNFSM4H5BWKTQ>
> .
>

lindsayad commented:
Closing this; I've got a ticket open for it
